File: core/src/main/java/com/uiptv/service/PlayerRequestResolver.java
package com.uiptv.service;

import com.uiptv.db.ChannelDb;
import com.uiptv.db.SeriesCategoryDb;
import com.uiptv.db.VodChannelDb;
import com.uiptv.model.*;
import com.uiptv.shared.Episode;

import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;

import static com.uiptv.util.StringUtils.isBlank;
import static com.uiptv.util.StringUtils.isNotBlank;

public class PlayerRequestResolver {

    public PlayerResponse resolveBookmarkPlayback(String bookmarkId, String mode, String seriesParentId) throws IOException {
        Bookmark bookmark = BookmarkService.getInstance().getBookmark(bookmarkId);
        Account account = AccountService.getInstance().getAll().get(bookmark.getAccountName());
        Account.AccountAction action = bookmark.getAccountAction() == null
                ? resolveMode(mode, account == null ? null : account.getAction())
                : bookmark.getAccountAction();
        Account playbackAccount = accountWithMode(account, action);
        Channel channel = resolveBookmarkChannel(bookmark);
        String scopedCategoryId = resolveSeriesCategoryId(playbackAccount, bookmark.getCategoryId());
        return PlayerService.getInstance().get(playbackAccount, channel, bookmark.getChannelId(), seriesParentId, scopedCategoryId);
    }

    public PlayerResponse resolveDirectPlayback(Account account, String categoryId, String channelId,
                                                String mode, String seriesParentId, String seriesId,
                                                Channel requestChannel) throws IOException {
        Account playbackAccount = accountWithMode(account, resolveMode(mode, account == null ? null : account.getAction()));
        Channel channel = mergeRequestChannel(resolveRequestedChannel(playbackAccount, categoryId, channelId, mode), requestChannel);
        String scopedCategoryId = resolveSeriesCategoryId(playbackAccount, categoryId);
        return PlayerService.getInstance().get(playbackAccount, channel, seriesId, seriesParentId, scopedCategoryId);
    }

    Channel resolveBookmarkChannel(Bookmark bookmark) {
        Channel channel = readBookmarkSnapshot(bookmark);
        if (channel != null) {
            return channel;
        }
        return createLegacyBookmarkChannel(bookmark);
    }

    Channel mergeRequestChannel(Channel channel, Channel requestChannel) {
        if (requestChannel == null) {
            return channel;
        }
        if (channel == null) {
            return requestChannel;
        }
        fillIfBlank(channel::getName, channel::setName, requestChannel.getName());
        fillIfBlank(channel::getLogo, channel::setLogo, requestChannel.getLogo());
        fillIfBlank(channel::getCmd, channel::setCmd, requestChannel.getCmd());
        fillIfBlank(channel::getCmd_1, channel::setCmd_1, requestChannel.getCmd_1());
        fillIfBlank(channel::getCmd_2, channel::setCmd_2, requestChannel.getCmd_2());
        fillIfBlank(channel::getCmd_3, channel::setCmd_3, requestChannel.getCmd_3());
        fillIfBlank(channel::getDrmType, channel::setDrmType, requestChannel.getDrmType());
        fillIfBlank(channel::getDrmLicenseUrl, channel::setDrmLicenseUrl, requestChannel.getDrmLicenseUrl());
        fillIfBlank(channel::getClearKeysJson, channel::setClearKeysJson, requestChannel.getClearKeysJson());
        fillIfBlank(channel::getInputstreamaddon, channel::setInputstreamaddon, requestChannel.getInputstreamaddon());
        fillIfBlank(channel::getManifestType, channel::setManifestType, requestChannel.getManifestType());
        fillIfBlank(channel::getSeason, channel::setSeason, requestChannel.getSeason());
        fillIfBlank(channel::getEpisodeNum, channel::setEpisodeNum, requestChannel.getEpisodeNum());
        return channel;
    }

    String resolveSeriesCategoryId(Account account, String rawCategoryId) {
        if (account == null || account.getAction() != Account.AccountAction.series) {
            return "";
        }
        if (isBlank(rawCategoryId)) {
            return "";
        }
        Category category = SeriesCategoryDb.get().getById(rawCategoryId);
        if (category != null && isNotBlank(category.getCategoryId())) {
            return category.getCategoryId();
        }
        return rawCategoryId;
    }

    private Channel readBookmarkSnapshot(Bookmark bookmark) {
        if (bookmark == null) {
            return null;
        }
        if (isNotBlank(bookmark.getSeriesJson())) {
            Episode episode = Episode.fromJson(bookmark.getSeriesJson());
            if (episode != null) {
                Channel channel = new Channel();
                channel.setCmd(episode.getCmd());
                channel.setName(episode.getTitle());
                channel.setChannelId(episode.getId());
                if (episode.getInfo() != null) {
                    channel.setLogo(episode.getInfo().getMovieImage());
                }
                return channel;
            }
        }
        if (isNotBlank(bookmark.getChannelJson())) {
            return Channel.fromJson(bookmark.getChannelJson());
        }
        if (isNotBlank(bookmark.getVodJson())) {
            return Channel.fromJson(bookmark.getVodJson());
        }
        return null;
    }

    private Channel createLegacyBookmarkChannel(Bookmark bookmark) {
        Channel channel = new Channel();
        channel.setCmd(decodeBookmarkCmd(bookmark.getCmd()));
        channel.setChannelId(bookmark.getChannelId());
        channel.setName(bookmark.getChannelName());
        channel.setDrmType(bookmark.getDrmType());
        channel.setDrmLicenseUrl(bookmark.getDrmLicenseUrl());
        channel.setClearKeysJson(bookmark.getClearKeysJson());
        channel.setInputstreamaddon(bookmark.getInputstreamaddon());
        channel.setManifestType(bookmark.getManifestType());
        return channel;
    }

    private String decodeBookmarkCmd(String value) {
        if (isBlank(value)) {
            return value;
        }
        try {
            return URLDecoder.decode(value, StandardCharsets.UTF_8);
        } catch (Exception _) {
            return value;
        }
    }

    private Channel resolveRequestedChannel(Account account, String categoryId, String channelId, String mode) {
        String normalizedMode = isBlank(mode) ? "" : mode.trim().toLowerCase();
        if ("vod".equals(normalizedMode) && account != null) {
            Channel vodChannel = VodChannelDb.get().getChannelByChannelId(channelId, categoryId, account.getDbId());
            if (vodChannel != null) {
                return vodChannel;
            }
            return VodChannelDb.get().getChannelByChannelIdAndAccount(channelId, account.getDbId());
        }
        return ChannelDb.get().getChannelById(channelId, categoryId);
    }

    private Account accountWithMode(Account account, Account.AccountAction action) {
        AccountMediaContext context = AccountMediaContext.from(account, action);
        return context == null ? null : context.toAccount();
    }

    private Account.AccountAction resolveMode(String mode, Account.AccountAction fallback) {
        if (isBlank(mode)) {
            return fallback == null ? Account.AccountAction.itv : fallback;
        }
        try {
            return Account.AccountAction.valueOf(mode.toLowerCase());
        } catch (Exception _) {
            return Account.AccountAction.itv;
        }
    }

    private void fillIfBlank(java.util.function.Supplier<String> getter,
                             java.util.function.Consumer<String> setter,
                             String value) {
        if (isBlank(getter.get()) && isNotBlank(value)) {
            setter.accept(value);
        }
    }
}

File: server/src/main/kotlin/com/uiptv/service/PlayerRequestResolver.kt
package com.uiptv.service

import com.uiptv.db.ChannelDb
import com.uiptv.db.SeriesCategoryDb
import com.uiptv.db.VodChannelDb
import com.uiptv.model.Account
import com.uiptv.model.Bookmark
import com.uiptv.model.Category
import com.uiptv.model.Channel
import com.uiptv.model.PlayerResponse
import com.uiptv.shared.Episode
import com.uiptv.util.StringUtils.isBlank
import com.uiptv.util.StringUtils.isNotBlank
import java.io.IOException
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.function.Consumer
import java.util.function.Supplier

class PlayerRequestResolver @JvmOverloads constructor(
    private val bookmarkService: BookmarkService = RuntimeServices.bookmarkService,
    private val accountService: AccountService = RuntimeServices.accountService,
    private val playerService: PlayerService = RuntimeServices.playerService,
    private val seriesCategoryDb: SeriesCategoryDb = SeriesCategoryDb.get(),
    private val vodChannelDb: VodChannelDb = VodChannelDb.get(),
    private val channelDb: ChannelDb = ChannelDb.get()
) {
    @Throws(IOException::class)
    fun resolveBookmarkPlayback(bookmarkId: String, mode: String?, seriesParentId: String?): PlayerResponse {
        val bookmark = bookmarkService.getBookmark(bookmarkId) ?: throw IOException("Bookmark not found")
        val account = accountService.getAll()[bookmark.accountName] ?: throw IOException("Account not found")
        applyMode(account, mode)
        bookmark.accountAction?.let { account.action = it }
        val channel = resolveBookmarkChannel(bookmark)
        val scopedCategoryId = resolveSeriesCategoryId(account, bookmark.categoryId)
        return playerService.get(account, channel, bookmark.channelId, seriesParentId, scopedCategoryId)
    }

    @Throws(IOException::class)
    fun resolveDirectPlayback(account: Account?, categoryId: String?, channelId: String?, mode: String?, seriesParentId: String?, seriesId: String?, requestChannel: Channel?): PlayerResponse {
        if (account == null) {
            throw IOException("Account not found")
        }
        applyMode(account, mode)
        val channel = mergeRequestChannel(resolveRequestedChannel(account, categoryId, channelId, mode), requestChannel)
        val scopedCategoryId = resolveSeriesCategoryId(account, categoryId)
        return playerService.get(account, channel, seriesId, seriesParentId, scopedCategoryId)
    }

    fun resolveBookmarkChannel(bookmark: Bookmark): Channel {
        return readBookmarkSnapshot(bookmark) ?: createLegacyBookmarkChannel(bookmark)
    }

    fun mergeRequestChannel(channel: Channel?, requestChannel: Channel?): Channel {
        if (requestChannel == null) return channel ?: Channel()
        if (channel == null) return requestChannel
        fillIfBlank({ channel.name }, { channel.name = it }, requestChannel.name)
        fillIfBlank({ channel.logo }, { channel.logo = it }, requestChannel.logo)
        fillIfBlank({ channel.cmd }, { channel.cmd = it }, requestChannel.cmd)
        fillIfBlank({ channel.cmd_1 }, { channel.cmd_1 = it }, requestChannel.cmd_1)
        fillIfBlank({ channel.cmd_2 }, { channel.cmd_2 = it }, requestChannel.cmd_2)
        fillIfBlank({ channel.cmd_3 }, { channel.cmd_3 = it }, requestChannel.cmd_3)
        fillIfBlank({ channel.drmType }, { channel.drmType = it }, requestChannel.drmType)
        fillIfBlank({ channel.drmLicenseUrl }, { channel.drmLicenseUrl = it }, requestChannel.drmLicenseUrl)
        fillIfBlank({ channel.clearKeysJson }, { channel.clearKeysJson = it }, requestChannel.clearKeysJson)
        fillIfBlank({ channel.inputstreamaddon }, { channel.inputstreamaddon = it }, requestChannel.inputstreamaddon)
        fillIfBlank({ channel.manifestType }, { channel.manifestType = it }, requestChannel.manifestType)
        fillIfBlank({ channel.season }, { channel.season = it }, requestChannel.season)
        fillIfBlank({ channel.episodeNum }, { channel.episodeNum = it }, requestChannel.episodeNum)
        return channel
    }

    fun resolveSeriesCategoryId(account: Account?, rawCategoryId: String?): String {
        if (account == null || account.action != Account.AccountAction.series) return ""
        if (isBlank(rawCategoryId)) return ""
        val category: Category? = seriesCategoryDb.getById(rawCategoryId.orEmpty())
        if (category != null && isNotBlank(category.categoryId)) return category.categoryId.orEmpty()
        return rawCategoryId.orEmpty()
    }

    private fun readBookmarkSnapshot(bookmark: Bookmark?): Channel? {
        if (bookmark == null) return null
        if (isNotBlank(bookmark.seriesJson)) {
            val episode = Episode.fromJson(bookmark.seriesJson)
            if (episode != null) {
                val channel = Channel()
                channel.cmd = episode.cmd
                channel.name = episode.title
                channel.channelId = episode.id
                channel.logo = episode.info?.movieImage
                return channel
            }
        }
        if (isNotBlank(bookmark.channelJson)) return Channel.fromJson(bookmark.channelJson.orEmpty())
        if (isNotBlank(bookmark.vodJson)) return Channel.fromJson(bookmark.vodJson.orEmpty())
        return null
    }

    private fun createLegacyBookmarkChannel(bookmark: Bookmark): Channel {
        val channel = Channel()
        channel.cmd = decodeBookmarkCmd(bookmark.cmd)
        channel.channelId = bookmark.channelId
        channel.name = bookmark.channelName
        channel.drmType = bookmark.drmType
        channel.drmLicenseUrl = bookmark.drmLicenseUrl
        channel.clearKeysJson = bookmark.clearKeysJson
        channel.inputstreamaddon = bookmark.inputstreamaddon
        channel.manifestType = bookmark.manifestType
        return channel
    }

    private fun decodeBookmarkCmd(value: String?): String? {
        if (isBlank(value)) return value
        return try {
            URLDecoder.decode(value, StandardCharsets.UTF_8)
        } catch (_: Exception) {
            value
        }
    }

    private fun resolveRequestedChannel(account: Account?, categoryId: String?, channelId: String?, mode: String?): Channel? {
        val normalizedMode = if (isBlank(mode)) "" else mode!!.trim().lowercase()
        if (normalizedMode == "vod" && account != null) {
            val vodChannel = vodChannelDb.getChannelByChannelId(channelId.orEmpty(), categoryId.orEmpty(), account.dbId.orEmpty())
            if (vodChannel != null) return vodChannel
            return vodChannelDb.getChannelByChannelIdAndAccount(channelId.orEmpty(), account.dbId.orEmpty())
        }
        return channelDb.getChannelById(channelId.orEmpty(), categoryId.orEmpty())
    }

    private fun applyMode(account: Account?, mode: String?) {
        if (account == null || isBlank(mode)) return
        try {
            account.action = Account.AccountAction.valueOf(mode!!.lowercase())
        } catch (_: Exception) {
            account.action = Account.AccountAction.itv
        }
    }

    private fun fillIfBlank(getter: Supplier<String?>, setter: Consumer<String?>, value: String?) {
        if (isBlank(getter.get()) && isNotBlank(value)) setter.accept(value)
    }
}
